examples/ethereum-tracker/src/contract.rs

```diff
@@ -69,69 +75,96 @@ impl Contract for EthereumTrackerContract {
+            .expect("Balance could not be parsed as a hex string");
+
+        self.state
+            .accounts
+            .insert(address, balance.into())
+            .expect("Failed to insert initial balance");
     }
 
-    async fn update(&mut self, to_block: u64) {
-        let event_name_expanded = "Transfer(address indexed,address indexed,uint256)";
-        let (ethereum_client, contract_address) = self.get_endpoints();
-        let start_block = self.state.start_block.get_mut();
-        let events = ethereum_client
-            .read_events(
-                &contract_address,
-                event_name_expanded,
-                *start_block,
-                to_block,
-            )
-            .await
-            .expect("Read a transfer event");
-        *start_block = to_block;
-        for event in events {
-            let EthereumDataType::Address(from) = event.values[0].clone() else {
-                panic!("wrong type for the first entry");
+    /// Updates the accounts based on the transfer events emitted up to the `end_block`.
+    async fn update(&mut self, end_block: u64) {
+        let request = async_graphql::Request::new(format!(
+            r#"query {{ readTransferEvents(endBlock: {end_block}) }}"#
+        ));
+
+        let application_id = self.runtime.application_id();
+        let response = self.runtime.query_service(application_id, request);
+
+        let async_graphql::Value::Object(data_object) = response.data else {
+            panic!("Unexpected response from `readTransferEvents`: {response:#?}");
+        };
+        let async_graphql::Value::List(ref events) = data_object["readTransferEvents"] else {
+            panic!("Unexpected response data from `readTransferEvents`: {data_object:#?}");
+        };
+
+        for event_value in events {
+            let async_graphql::Value::Object(event) = event_value else {
+                panic!("Unexpected event returned from `readTransferEvents`: {event_value:#?}");
             };
-            let EthereumDataType::Address(to) = event.values[1].clone() else {
-                panic!("wrong type for the second entry");
+
+            let async_graphql::Value::String(ref source) = event["source"] else {
+                panic!("Unexpected source address in transfer event: {event:#?}");
+            };
+            let async_graphql::Value::String(ref destination) = event["destination"] else {
+                panic!("Unexpected destination address in transfer event: {event:#?}");
             };
-            let EthereumDataType::Uint256(value) = event.values[2] else {
-                panic!("wrong type for the third entry");
+            let async_graphql::Value::String(ref value_string) = event["value"] else {
+                panic!("Unexpected balance in transfer event: {event:#?}");
             };
+
+            let value = U256::from_str_radix(&value_string[2..], 16)
+                .expect("Balance could not be parsed as a hex string");
```

Comment:
It is a little unfortunate to have this rather low level conversion occurring in this part of the code.

Reply:
I'd imagine that `FromStr` would work on the whole string here, and interpret it as hexadecimal because it starts with `0x`? I.e. maybe we can just use `value_string.parse::<U256>()`?

Reply:
I still see `from_str_radix`.